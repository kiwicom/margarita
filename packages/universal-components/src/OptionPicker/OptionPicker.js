// @flow

import * as React from 'react';
import { View } from 'react-native';
import { defaultTokens } from '@kiwicom/orbit-design-tokens';

import type { OptionTypeInterface } from './OptionPickerTypes';
import { StyleSheet } from '../PlatformStyleSheet';
import { TagsInput } from '../TagsInput';
import OptionList from './components/OptionList';

type Props = {|
  +onChangeText: string => void,
  +onClearPress: () => void,
  +onPressAdd: OptionTypeInterface => void,
  +onPressItem: OptionTypeInterface => void,
  +text?: string,
  +label?: string,
  +onKeyPress?: Event => void,
  +options?: Array<OptionTypeInterface>,
  +placeholder?: string,
  +selected?: ?Array<OptionTypeInterface>,
|};

type Event = { nativeEvent: { key: string } };

export default function OptionPicker(props: Props) {
  const {
    options,
    label,
    placeholder,
    text,
    selected,
    onClearPress,
    onChangeText,
    onKeyPress,
    onPressItem,
    onPressAdd,
  } = props;

  const tags = selected ? selected.map(option => option.label) : [];

  return (
    <View>
      <View style={styles.inputWrapper}>
        <TagsInput
          autoFocus
          tags={tags}
          onClearPress={onClearPress}
          value={text}
          onChangeText={onChangeText}
          onKeyPress={onKeyPress}
          label={label}
          placeholder={placeholder}
        />
      </View>
      {options && (
        <OptionList
          onItemPress={onPressItem}
          onAddPress={onPressAdd}
          options={options}
        />
      )}
    </View>
  );
}

const commonStyles = {
  borderBottomWidth: 1,
  borderBottomColor: defaultTokens.borderColorInput,
};

const styles = StyleSheet.create({
  inputWrapper: {
    padding: 5,
    ios: {
      ...commonStyles,
    },
    android: {
      ...commonStyles,
    },
  },
});
