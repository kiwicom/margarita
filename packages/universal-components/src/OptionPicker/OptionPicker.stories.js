// @flow

import * as React from 'react';
import { storiesOf } from '@storybook/react-native';
import { withKnobs, text, select, object } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';

import mockedPlaces, { Prague } from './mocks/places';

import { OptionPicker } from '.';

storiesOf('OptionPicker', module)
  .addDecorator(getStory => getStory())
  .addDecorator(withKnobs)
  .add('Playground', () => {
    const onChangeText = action('onChangeText');
    const onClearPress = action('onClearPress');
    const onPressAdd = action('onPressAdd');
    const onPressItem = action('onPressItem');
    const selected = select(
      'selected',
      {
        None: null,
        Prague: [Prague],
      },
      null,
    );
    const options = object('Options', mockedPlaces);
    const label = text('label', 'From:');
    const placeholder = text('placeholder', 'Departure point');

    return (
      <OptionPicker
        selected={selected}
        onClearPress={onClearPress}
        label={label}
        placeholder={placeholder}
        onChangeText={onChangeText}
        onPressAdd={onPressAdd}
        onPressItem={onPressItem}
        options={options}
      />
    );
  });
