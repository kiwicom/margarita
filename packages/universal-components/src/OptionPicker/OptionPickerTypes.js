// @flow

export interface OptionTypeInterface {
  +id: string | number;
  +label: string;
  +text?: string;
  +type: 'localization' | 'destination' | 'airplane' | 'bus' | 'train'; // @TODO use IconNameType as enum types
  +subOptions?: Array<OptionTypeInterface>;
}
