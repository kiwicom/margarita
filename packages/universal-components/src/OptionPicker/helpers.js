// @flow

import type { OptionTypeInterface } from './OptionPickerTypes';

export function getOptionsIds(options: OptionTypeInterface[]) {
  return options.reduce((result, option) => {
    let ids = [option.id];
    if (option.subOptions) {
      const subOptions = option.subOptions.map(({ id }) => id);
      ids = [...ids, ...subOptions];
    }

    return [...result, ...ids];
  }, []);
}

export function areSelectedOptionsChanged(
  optionsA: OptionTypeInterface[],
  optionsB: OptionTypeInterface[],
) {
  const optionsIdsB = getOptionsIds(optionsB);
  const optionsIdsA = getOptionsIds(optionsA);

  return (
    optionsIdsA.length !== optionsIdsB.length ||
    !optionsIdsA.every(option => optionsIdsB.includes(option))
  );
}
