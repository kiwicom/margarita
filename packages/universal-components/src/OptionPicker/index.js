// @flow

export { default as OptionPicker } from './OptionPicker';
