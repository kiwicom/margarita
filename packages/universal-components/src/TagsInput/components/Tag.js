// @flow

import * as React from 'react';
import { View } from 'react-native';
import { defaultTokens } from '@kiwicom/orbit-design-tokens';

import { Text } from '../../Text';
import { StyleSheet, type StylePropType } from '../../PlatformStyleSheet';
import { designTokens } from '../../DesignTokens';
import { Icon } from '../../Icon';
import { ExtendedTouchable } from '../../ExtendedTouchable';

type Props = {|
  +children: React.Node,
  +style?: StylePropType,
  +testID?: string,
  +index: number,
  +onDeletePress?: number => void,
  +fontSize: number,
|};

type TouchableTagIconProps = {|
  +onPress?: number => void,
  +children: React.Node,
  +index: number,
|};

class TouchableTagIcon extends React.Component<TouchableTagIconProps> {
  onPress = () => {
    if (this.props.onPress) this.props.onPress(this.props.index);
  };

  render() {
    return this.props.onPress ? (
      <ExtendedTouchable onPress={this.onPress}>
        {this.props.children}
      </ExtendedTouchable>
    ) : (
      this.props.children
    );
  }
}
export default function Tag({
  children,
  style,
  testID,
  index,
  onDeletePress,
  fontSize,
}: Props) {
  const dynamicStyle = StyleSheet.create({ fontSize: { fontSize: fontSize } });

  return (
    <View style={[styles.wrapper, style]} testID={testID}>
      <TouchableTagIcon onPress={onDeletePress} index={index}>
        <Icon
          name="close-circle"
          style={styles.icon}
          size="small"
          color={defaultTokens.backgroundButtonWhite}
        />
      </TouchableTagIcon>
      <Text style={[styles.text, dynamicStyle.fontSize]}>{children}</Text>
    </View>
  );
}

Tag.defaultProps = {
  fontSize: parseFloat(defaultTokens.fontSizeTextLarge),
};

const styles = StyleSheet.create({
  wrapper: {
    paddingHorizontal: 6,
    paddingVertical: 5,
    alignSelf: 'flex-start',
    flexDirection: 'row',
    alignItems: 'center',
    flexWrap: 'wrap',
    borderRadius: parseFloat(defaultTokens.borderRadiusNormal),
    backgroundColor: defaultTokens.backgroundButtonPrimary,
    web: {
      width: 'fit-content',
    },
  },
  text: {
    lineHeight: parseFloat(designTokens.heightBadge),
    letterSpacing: 0.3,
    color: defaultTokens.colorTextBadgeDark,
    marginStart: 5,
  },
  icon: {
    opacity: 0.6,
  },
});
