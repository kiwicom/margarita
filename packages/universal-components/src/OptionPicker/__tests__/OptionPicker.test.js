// @flow

import React from 'react';
import { TextInput } from 'react-native';
import { shallow, render, fireEvent } from 'react-native-testing-library';

import OptionPicker from '../OptionPicker';
import options, { Prague, Berlin } from '../mocks/places';
import Tag from '../../TagsInput/components/Tag';

jest.mock('NativeAnimatedHelper');

describe('OptionPicker', () => {
  const label = 'label';
  const placeholder = 'placeholder';
  const onChangeText = jest.fn();
  const onPressAdd = jest.fn();
  const onPressItem = jest.fn();
  const onKeyPress = jest.fn();
  const onClearPress = jest.fn();
  const selected = [Berlin];
  const value = Berlin.label;

  const {
    getByText,
    getByProps,
    getByTestId,
    getAllByType,
    getByType,
  } = render(
    <OptionPicker
      onClearPress={onClearPress}
      onKeyPress={onKeyPress}
      text={value}
      label={label}
      selected={selected}
      placeholder={placeholder}
      onChangeText={onChangeText}
      options={options}
      onPressAdd={onPressAdd}
      onPressItem={onPressItem}
    />,
  );
  test('passed the props', () => {
    expect(
      getByProps({
        selected,
        onChangeText,
        options,
        onPressAdd,
        onPressItem,
        onClearPress,
        onKeyPress,
        label,
        placeholder,
      }),
    ).toBeDefined();
    expect(getByText(label)).toBeDefined();
    expect(getAllByType(Tag)).toHaveLength(1);
  });

  test('call the onPressItem callback', () => {
    const Item = getByText(Prague.label);
    fireEvent(Item, 'press');
    expect(onPressItem).toHaveBeenCalledWith(Prague);
  });

  test('call the onClearPress callback', () => {
    const Item = getByTestId('delete-button');
    fireEvent(Item, 'press');
    expect(onClearPress).toHaveBeenCalledTimes(1);
  });

  test('call the onChangeText callback', () => {
    fireEvent.changeText(getByType(TextInput), 'CHANGE_TEXT');
    expect(onChangeText).toHaveBeenLastCalledWith('CHANGE_TEXT');
  });

  test('render', () => {
    const { output } = shallow(
      <OptionPicker
        onClearPress={onClearPress}
        onKeyPress={onKeyPress}
        text={value}
        label={label}
        selected={selected}
        placeholder={placeholder}
        onChangeText={onChangeText}
        options={options}
        onPressAdd={onPressAdd}
        onPressItem={onPressItem}
      />,
    );

    expect(output).toMatchSnapshot();
  });

  test('should display placeholder', () => {
    const { getByPlaceholder } = render(
      <OptionPicker
        text=""
        onClearPress={onClearPress}
        onKeyPress={onKeyPress}
        label={label}
        placeholder={placeholder}
        onChangeText={onChangeText}
        options={options}
        onPressAdd={onPressAdd}
        onPressItem={onPressItem}
      />,
    );
    expect(getByPlaceholder(placeholder)).toBeDefined();
  });
});
