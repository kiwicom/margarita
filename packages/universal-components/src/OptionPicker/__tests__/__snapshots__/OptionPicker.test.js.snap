// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionPicker render 1`] = `
<Component>
  <Component
    style={
      Object {
        "borderBottomColor": "#bac7d5",
        "borderBottomWidth": 1,
        "padding": 5,
      }
    }
  >
    <TagsInput
      autoCorrect={false}
      autoFocus={true}
      fontSize={16}
      label="label"
      onChangeText={
        [MockFunction] {
          "calls": Array [
            Array [
              "CHANGE_TEXT",
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        }
      }
      onClearPress={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        }
      }
      onKeyPress={[MockFunction]}
      placeholder="placeholder"
      tags={
        Array [
          "Berlin",
        ]
      }
      value="Berlin"
    />
  </Component>
  <OptionList
    onAddPress={[MockFunction]}
    onItemPress={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "id": "prague",
              "label": "Prague",
              "subOptions": Array [
                Object {
                  "id": "train-station",
                  "label": "Prague Central Station",
                  "text": "5 km from city center",
                  "type": "train",
                },
                Object {
                  "id": "airport",
                  "label": "Václav Havel Airport",
                  "text": "5 km from city center",
                  "type": "airplane",
                },
              ],
              "text": "Czech Republic",
              "type": "destination",
            },
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    options={
      Array [
        Object {
          "id": "prague",
          "label": "Prague",
          "subOptions": Array [
            Object {
              "id": "train-station",
              "label": "Prague Central Station",
              "text": "5 km from city center",
              "type": "train",
            },
            Object {
              "id": "airport",
              "label": "Václav Havel Airport",
              "text": "5 km from city center",
              "type": "airplane",
            },
          ],
          "text": "Czech Republic",
          "type": "destination",
        },
        Object {
          "id": "berlin",
          "label": "Berlin",
          "subOptions": Array [
            Object {
              "id": "QPP",
              "label": "Airport of Berlin",
              "text": "4.5 km from city center",
              "type": "airplane",
            },
            Object {
              "id": "QWC",
              "label": "Zoo Railway Station",
              "text": "10 km from city center",
              "type": "airplane",
            },
          ],
          "text": "Federal Republic of Germany",
          "type": "destination",
        },
        Object {
          "id": "presov",
          "label": "Presov",
          "text": "Slovakia",
          "type": "destination",
        },
      ]
    }
  />
</Component>
`;
